aten/src/ATen/native/WeightNorm.cpp

```diff
@@ -0,0 +1,75 @@
+#include "ATen/ATen.h"
+#include "ATen/TensorUtils.h"
+#include "ATen/NativeFunctions.h"
+
+#include <cstring>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+
+namespace at { 
+namespace native {
+
+// Staying faithful to the Python for now for clarity, look for optimizations later
+// (eg single return statement for RVO)
+Tensor norm_except_dim(const Tensor & v, int64_t pow, int64_t dim)
+{
+  // I assume tensor.contiguous(), view(), norm(), etc. here will dispatch through VariableType.
+  if(dim == -1)
+    return v.norm(pow);
+  else if(dim == 0)
+  {
+    std::vector<int64_t> output_size(v.dim(), 1);
+    output_size[0] = v.size(0);
+    return v.contiguous().view({v.size(0), -1}).norm(pow, 1).view(output_size);
```

Comment:
suggestion: use reshape here instead of contiguous + view.

Reply:
Thank you for the quick review!  I had a couple reasons for choosing this pattern:
1.  I'm mimicking the original Python:  https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/weight_norm.py#L13
2.  norm_except_dim is accessible from the Python, and my implement [uses it](https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/weight_norm.py#L13)) on the Python side, not just in _weight_norm on the C++ side.  IMO it's also a useful independent Python function to have.  For these two reasons, in general it's possible the incoming v may be noncontiguous, have pathologically permuted indices, etc.  I need the tensor's raw data to be C-style contiguous, so that A. it can be viewed with all but one dimension flattened and B.  the norm will be performant.  I'm not sure if reshape does all these things, so I decided to stick with what I knew would work.
3.  The ops in here need to be (double)differentiable.  I'm also not sure how reshape interacts with autograd (I expect it's fine and i'm being paranoid).

If 2. and 3. are not actually problems with reshape, I can use reshape instead.

Reply:
The subsequent norm op involves (at least) 3 tensor-size round trips to global memory for a non-persistent implementation (2 reads and a write), so it's worth making v C-style contiguous for coalesced memory accesses during the norm.  Unfortunately reshape() is not strict enough to guarantee C-style contiguity in all cases:
```
>>> a = torch.Tensor([[1,2],[3,4]])
>>> a.t().reshape(a.size(0), -1).stride()
(1, 2)
>>> a.t().contiguous().view(a.size(0), -1).stride()
(2, 1)
```